### PR TITLE
networkd link serialization fix

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/libsystemd-network/network-internal.c [hunk 1]
```diff
@@ -772,16 +769,14 @@ int deserialize_in6_addrs(struct in6_addr **ret, const char *string) {
...
-
         assert(f);
         assert(key);
         assert(routes);
         assert(size);
 
         fprintf(f, "%s=", key);
 
-        for (i = 0; i < size; i++) {
+        for (unsigned i = 0; i < size; i++) {
```
Comment (poettering) [inline src/libsystemd-network/network-internal.c]: given that `size` is of `size_t` (which I think is the right type) `i` should probably be too

File: src/network/networkd-link.c [hunk 2]
```diff
@@ -4119,100 +4150,50 @@ int link_save(Link *link) {
...
+                                    SD_DHCP_LEASE_SMTP_SERVERS,
+                                    true,
+                                    NULL, NULL);
+
+                serialize_addresses(f, "LPR_SERVERS", NULL,
+                                    link->network->lpr,
+                                    link->dhcp_lease,
+                                    SD_DHCP_LEASE_LPR_SERVERS,
+                                    true,
+                                    NULL, NULL);
```
Comment (poettering) [inline src/network/networkd-link.c]: hmm, I think we should probably drop the _SERVERS suffix for this too, no? i.e. DNS and SIP and NTP don't have such a suffix, and these are just the same

**Comment by poettering** [comment]
CI appears unhappy?

Maybe @yuwata/@ssahani should have a look too

File: src/network/networkd-link.c (hunk 2 above)
Comment (keszybz) [reply src/network/networkd-link.c]: Yeah, I think we should do that, but that is mostly cosmetic. I would prefer to figure out the part I added to TODO first, i.e. how to actually test this. Apparently this wasn't tested at all, and various link files that I have on running machines only have a small subset of fields defined, so it's hard to test serialization comprehensively.

**Comment by mrc0mmand** [comment]
There's actually quite lot going on:

1)
```
systemd-networkd[124044]: DHCPv6 CLIENT: T1 expires in 27min 53s
systemd-networkd[124044]: DHCPv6 CLIENT: T2 expires in 49min 5s
systemd-networkd[124044]: veth99: DHCPv6 address 2600::1d/128 timeout preferred 3600 valid 3600
systemd-networkd[124044]: DHCPv6 CLIENT: Recv REPLY
systemd-networkd[124044]: Assertion 'size' failed at src/libsystemd-network/network-internal.c:723, function void serialize_in6_addrs(FILE *, const struct in6_addr *, size_t)(). Aborting.
```

2)
```
systemd-networkd[175653]: DHCPv6 CLIENT: Sent SOLICIT
systemd-networkd[175653]: DHCPv6 CLIENT: Next retransmission in 1s
systemd-networkd[175653]: veth99: Received remembered route: dst: n/a, src: n/a, gw: fe80::1034:56ff:fe78:9abd, prefsrc: n/a, scope: global, table: main, proto: ra, type: unicast
systemd-networkd[175653]: veth99: Received remembered route: dst: 2600::/64, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: ra, type: unicast
systemd-networkd[175653]: DHCPv6 CLIENT: T1 expires in 29min 56s
systemd-networkd[175653]: DHCPv6 CLIENT: T2 expires in 50min 23s
systemd-networkd[175653]: veth99: DHCPv6 address 2600::10/128 timeout preferred 3600 valid 3600
systemd-networkd[175653]: DHCPv6 CLIENT: Recv REPLY
systemd-networkd-wait-online[175664]: veth99: link is being processed by networkd
systemd-resolved[172119]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/resolve1 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=1268 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
kernel: systemd-network[175653]: segfault at 1823000 ip 00007f9a64379028 sp 00007ffde23fbe40 error 4 in libc-2.17.so[7f9a6426c000+1c3000]
systemd[1]: Started Process Core Dump (PID 175666/UID 0).
systemd-udevd[172122]: Network interface NamePolicy= disabled on kernel command line, ignoring.
systemd-udevd[172122]: /usr/lib/udev/rules.d/11-dm-lvm.rules:40 Invalid value for OPTIONS key, ignoring: 'event_timeout=180'
systemd-udevd[172122]: /usr/lib/udev/rules.d/11-dm-lvm.rules:40 The line takes no effect, ignoring.
systemd-udevd[172122]: /usr/lib/udev/rules.d/84-nm-drivers.rules:10 Invalid value "/bin/sh -c 'ethtool -i $1 | sed -n s/^driver:\ //p' -- $env{INTERFACE}" for PROGRAM (char 24: invalid substitution type), ignoring, but please fix it.
systemd[1]: systemd-networkd.service: Main process exited, code=dumped, status=11/SEGV
systemd[1]: systemd-networkd.service: Failed with result 'core-dump'.
systemd-coredump[175667]: Process 175653 (systemd-network) of user 192 dumped core.
                          
                          Stack trace of thread 175653:
                          #0  0x00007f9a64379028 inet_ntop (libc.so.6 + 0x10d028)
                          #1  0x00000000004a29ba serialize_in6_addrs (systemd-networkd + 0xa29ba)
                          #2  0x000000000046977d serialize_addresses (systemd-networkd + 0x6977d)
                          #3  0x00000000004742d2 link_save (systemd-networkd + 0x742d2)
                          #4  0x0000000000413a57 manager_dirty_handler (systemd-networkd + 0x13a57)
                          #5  0x00007f9a64a16c43 source_dispatch (libsystemd-shared-245.so + 0x1c0c43)
                          #6  0x00007f9a64a18d2a sd_event_dispatch (libsystemd-shared-245.so + 0x1c2d2a)
                          #7  0x00007f9a64a18fbc sd_event_run (libsystemd-shared-245.so + 0x1c2fbc)
                          #8  0x00007f9a64a1912d sd_event_loop (libsystemd-shared-245.so + 0x1c312d)
                          #9  0x000000000040cca5 run (systemd-networkd + 0xcca5)
                          #10 0x000000000040cd69 main (systemd-networkd + 0xcd69)
                          #11 0x00007f9a6428e555 __libc_start_main (libc.so.6 + 0x22555)
                          #12 0x000000000040c3b9 _start (systemd-networkd + 0xc3b9)
systemd[1]: systemd-coredump@12-175666-0.service: Succeeded.
```

The full coredump with gdb backtrace: https://ci.centos.org/job/systemd-pr-build/12024/artifact//systemd-centos-ci/artifacts_W7gVqs/testsuite-logs-upstream.och/coredumpctl_collect_FAIL.log

The usefulness is questionable, though. Hopefully the KVM job will give better results once it finishes.

**Comment by keszybz** [comment]
Thanks @mrc0mmand. I completely forgot to pass the size properly. I reworked the code a bit and I think it is clearer now.

(I was considering whether to add a similar getter function for server lists for dhcp6, but looking at the option list for dhcp6, there doesn't seem to be that many other options with servers lists, ntp and dns are already the main ones. So I think it is OK to keep using individual getter functions.)

**Pull request merged**